Instrumentation Ecosystem Management (#155)
open-telemetry/oteps: text/0155-external-modules.md

> # Instrumentation Ecosystem Management

Reviewer: Should this guidance also apply to other ecosystem components like exporters, propagators, resource detectors, etc.? Those things are also in the registry/contrib repos.

Author: They should indeed. Do you have suggestions for better title?

Reviewer: I don't have any great suggestion, maybe just "Ecosystem Management"?

Author: changed name

Proposal how to leverage wider community contributing instrumentations and other packages.

## Motivation

> For OpenTelemetry to become a de-facto standard in observability there must exist a vast ecosystem of OpenTelemetry components.
> Including integrations with various libraries and frameworks in all languages supported by OpenTelemetry.

Reviewer 2: ```suggestion
For OpenTelemetry to become a de-facto standard in observability there must exist a vast ecosystem of OpenTelemetry components, 
including integrations with various libraries and frameworks in all languages supported by OpenTelemetry.
```

> We cannot possibly expect that all these integrations will be provided by the core maintainers of OpenTelemetry.

Reviewer 3: +1. We should encourage this as much as we can by providing great examples and documentation on how to create such integrations for various classes of software, by helping the third-party projects if they need help while they work on integrations.
This likely warrants a separate effort that is outside of this PR's scope.

> We cannot possibly expect that all these integrations will be provided by the core maintainers of OpenTelemetry.

Author: Should I include some words about this into this OTEP?

> We cannot possibly expect that all these integrations will be provided by the core maintainers of OpenTelemetry.

Reviewer 4: Please do!  I think at least calling out this goal is important (and a lot of compoennts of that are out of scope)

We have to have a way to leverage a wider community while still providing our end-users with some way to discover
all available Otel components together with some visibility into their quality.

## Explanation

> [OpenTelemetry Registry](https://opentelemetry.io/registry/) serves as a central catalogue of all known Otel components,

Reviewer 2: ```suggestion
The [OpenTelemetry Registry](https://opentelemetry.io/registry/) serves as a central catalogue of all known Otel components,
```

both provided by core maintainers of the project and any third party.

In order for a component to be included into Registry its authors have to fill a [self-assessment form](#registry-self-assessment-form).

> Registry should allow a clear visibility if component came from a repository in OpenTelemetry organisation.

Reviewer 2: ```suggestion
The registry should allow a clear visibility if component came from a repository in OpenTelemetry organisation.
```

> Registry should allow a clear visibility if component came from a repository in OpenTelemetry organisation.

Reviewer 5: I think having strong registry is critical to this OTEP, so it makes sense to pay more attention to it and list the requirements it must satisfy, i.e. what required information it must capture about each instrumentation component, e.g. things like
  * location url
  * language
  * target instrumented library
  * some compatibility info, like the versions of OTEL & the instrumented library
  * license
  * dimensions of the self-assessment form
  * official status (maintained by SIG | in contrib | 3rd party)
  * any quality indicators we could standardize on (e.g. test coverage)
  * "popularity" (e.g. github stars)
  * dev stats (date of creation, date of last release, date of last commit)

The website can have a daily cron job to rebuild the registry with fresh info

> Registry should allow a clear visibility if component came from a repository in OpenTelemetry organisation.

Author: Do you think this list of all information pieces should be agreed upon and finalized in this OTEP? Or this will be a part of the effort of implementing this OTEP in the Registry? I think the latter.

> Registry should allow a clear visibility if component came from a repository in OpenTelemetry organisation.

Reviewer 5: I think it should be part of this OTEP, since it's a list of requirements to the registry that *this* OTEP depends on (doesn't have to be all of the above, but there's a minimal required set).

> Registry should allow a clear visibility if component came from a repository in OpenTelemetry organisation.

Author: Added.

> A component can be removed from the Registry if any declaration from the self-assessment form is violated and not remedied

Author: Will add some words about this.

in a timely manner, provisionally one or two weeks.

## Internal details

We distinguish the following sources of OpenTelemetry components and integrations.

### Native or built-in instrumentations

> Any library or framework can use OpenTelemetry API to natively produce Telemetry.

Reviewer 2: ```suggestion
Any library or framework can use the OpenTelemetry API to natively produce telemetry.
```

> We encourage library authors to submit their library for inclusion into OpenTelemetry Registry.

Reviewer 2: ```suggestion
We encourage library authors to submit their library for inclusion into the OpenTelemetry Registry.
```

### Core components

> Otel SIGs may provide instrumentations for any libraries that are deemed important enough by SIG’s maintainers.

Reviewer 2: ```suggestion
Otel SIGs may provide instrumentation for any libraries that are deemed important enough by the SIG’s maintainers.
```

> By doing this the SIG maintainers commit to support these instrumentations (including future versions of the library),

Reviewer 2: ```suggestion
By doing this the SIG maintainers commit to support this instrumentation (including future versions of the library),
```

provide updates/fixes, including security patches and guarantee that they comply with Otel semantic conventions and best practices.

> Depending on the SIG these core instrumentations may share repository, infrastructure and maintainers with

Reviewer 2: ```suggestion
Depending on the SIG, these instances of core instrumentation may share the repository, infrastructure and maintainers with
```

> OpenTelemetry API/SDK implementation for this language or be separate.

Reviewer 2: ```suggestion
the OpenTelemetry API/SDK implementation for this language or be separate.
```

> All core instrumentations must be included into Otel Registry following the usual process described above.

Reviewer 2: ```suggestion
All instances of core instrumentation must be included into the Otel Registry following the usual process described above.
```

### Contrib components

Any language SIG may have one or more “contrib” repos containing components contributed by external developers.
This repository may have a separate set of approvers/maintainers than the core API/SDK repo.
SIG maintainers are encouraged to have a simplified process to promote contributors to approver/maintainer role in this repository.

> Contrib repository may leverage the CODEOWNERS functionality of GitHub to assign maintainers to individual packages

Reviewer 2: ```suggestion
A contrib repository may leverage the CODEOWNERS functionality of GitHub to assign maintainers to individual packages
```

even if this means granting write permissions to the whole repo.
The goal should be to distribute the load of reviewing PRs and accepting changes as much as possible.

> All components in contrib repository are expected to be included into Otel Registry following the usual process described above.

Reviewer 2: ```suggestion
All components in a contrib repository are expected to be included into the Otel Registry following the usual process described above.
```

> It should be very easy for external contributors to include their components into contrib repo as opposed to hosting them separately.

Reviewer 2: ```suggestion
It should be very easy for external contributors to include their components into a contrib repo as opposed to hosting them separately.
```

This way they can reuse existing infrastructure for testing, publishing, security scanning etc.

> Also this will greatly simplify responsibility transfer between different maintainers if their priorities change. It also promotes the development

Reviewer 2: ```suggestion
This will also greatly simplify responsibility transfer between different maintainers if their priorities change. It also promotes the development
```

> Also this will greatly simplify responsibility transfer between different maintainers if their priorities change. It also promotes the development

Reviewer 4: What's the story here of instrumentation that STARTS in openTelemetry and eventually becomes "native" (your first use case).

I think we need to (eventually) provide these kinds of utilities for that first party telemetry.    This makes me think we don't want to consider providing tools for making external telemetry easier to maintain for that use case.

Author: We absolutely should and this is implied by self-assessment form below, which is applicable to native instrumentations as well:

> If Otel/SIG provides a testing harness to verify produced telemetry, that test was used and passed

and maintenance of a single instrumentation package for each instrumentation source, so that work isn't spread amongst multiple parallel solutions.

Language SIGs are encouraged to provide a testing harness to verify that component adheres to OpenTelemetry semantic conventions

> and [recommendations for Otel instrumentations design](https://docs.google.com/document/d/1YNRCg9fdjJgZRs56vvf7rfFPk06mhp781sWHYypOaAk/edit#)

Reviewer 5: I am not seeing "instrumentation design" recommendation in that doc

Author: That document was created to oversee all the required work that has to be done to formulate OpenTelemetry recommendations for instrumentations and semantic convention usage. Unfortunately no such recommendation has been completed yet. My assumption/hope that it will be done still.

Author: Also, the sentence ends with

> **when** OpenTelemetry starts publishing them

Author: Link removed

when OpenTelemetry starts publishing them.

A high volume of contrib contributions presents a burden for language maintainers. There are two suggestions for tackling this:

> - Create an "experimental" folder within contrib. The contents of this folder are not reviewed or maintained by language repository maintainers, but

Reviewer 6: What about `alpha` instead? I'm guessing you mean they are still under development, rather than being a funky or exotic new feature ;)

Author: I will defer that to @mtwo as the original author of this suggestion.

Reviewer 6: @mtwo Ping ;)

Author: They should be added to the registry. I have added the requirement for labelling. 

@carlosalberto I don't have a strong preference about the name. I think we will have to somehow describe the status of contrib repositories somewhere. During that submission we may come up with the final name.

many of the other benefits of being within a contrib repository remain.
- Add more approvers and maintainers, perhaps some who exclusively focus on submissions to contrib.

### External components

> If component authors for whatever reason want to host their contribution outside the Otel contrib repository they are free to do so (though we

Reviewer 2: ```suggestion
If component authors, for whatever reason, want to host their contribution outside an Otel contrib repository they are free to do so (though we
```

encourage all contributions to go into contrib or core language repositories). Their submission for inclusion into OpenTelemetry Registry is still
welcomed, subject to the same process described above.

### Distribution

> Whenever OpenTelemetry components are published to any repository other than OpenTelemetry Registry (such as npm registry or Maven Central)

Reviewer 2: ```suggestion
Whenever OpenTelemetry components are published to any repository other than the OpenTelemetry Registry (such as npm registry or Maven Central)
```

only core and contrib components can be published under "opentelemetry" namespace.

> Native and external components are to be published under their own namespace.

Reviewer 3: What is a "native" component?

Author: Section `Native or built-in instrumentations` above.

> In case Otel SIG provides any kind of "all-in-one" instrumentation distribution (e.g. Java and .NET do) they should include

Reviewer 2: ```suggestion
In case the Otel SIG provides any kind of "all-in-one" instrumentation distribution (e.g. as Java and .NET do) they should include
```

> only core and contrib packages into it.

Reviewer 5: Why this restriction? It's akin to saying that Jaeger binary (for example) may only include code from Jaeger repos, no 3rd party dependencies. If there's a high quality instrumentation hosted by 3rd party, I don't see why the distro cannot include it.

Reviewer 4: +1.

Maybe be less restrictive.   

```
If Otel provides bundled instrumentation distributions ONE of the available options should only include instrumentation under the OpenTelemetry provenance.
```

For bundled instrumentation distributions WITH third party instrumentation, we need to make sure issues/support is directed to the appropriate location, but I think we can solve that issue.

Author: This restriction was included based on the feedback on the initial draft: https://docs.google.com/document/d/1YNRCg9fdjJgZRs56vvf7rfFPk06mhp781sWHYypOaAk/edit?disco=AAAAMSIJr4A

Reviewer 4: The concern was about security risk.   I don't think this clause is the only way to deal with security concerns, nor does it alleviate security risk in any bundled offering of -contrib.   I think we need more discussion around this limtiation, I'd be very against making it without more thought on security overall.

Author: I tried to modify wording, PTAL again.

> OpenTelemetry Registry should provide a way to easily obtain a list of these components.

Reviewer 2: ```suggestion
The OpenTelemetry Registry should provide a way to easily obtain a list of these components.
```

> If possible SIG should provide a mechanism to include any external component during target application's build- or runtime.

Reviewer 2: ```suggestion
If possible, SIGs should provide a mechanism to include any external component during target application's build- or runtime.
```

This may mean a separate language-specific component API that all components are encouraged to implement.

## Trade-offs and mitigations

* How easy it is to get merge permission in contrib repo?
The harder it is, the larger is the maintenance burden on the core team.

> The easier it is, the more uncertainty there is about the quality of contributions.

Reviewer 4: I think an easier problem to solve may be ensuring the quality of instrumentation is clearly visible to new users and folks trying new instrumentation can clearly communicate quality.   This has a "social pressure" to force good quality instrumentation vs. relying on our own governance model to handle all of that.

Author: Added this below into "Open questions" section

Can every language SIG decide this for themselves or should we decide together?

## Open questions

### Registry self-assessment form

> The exact list should be developed separately but at least component's author should declare that

Reviewer 2: ```suggestion
The exact list should be developed separately, but at least component's author should declare that
```

> * It uses permissive OSS license

Reviewer 5: Should we insist on no copyleft licenses, i.e. only allow licenses approved by CNCF?

Author: This is a good idea, I will add it clause.

* It does not have any known security vulnerabilities
* It produces telemetry which adheres to OpenTelemetry semantic conventions

> * If Otel/SIG provides a testing harness to verify produced telemetry, that test was used and passed

Reviewer 2: ```suggestion
* If the Otel/SIG provides a testing harness to verify produced telemetry, that tests were used and passed
```

We have to encourage that by providing great documentation, examples and tooling to integration authors,
while still providing our end-users with some way to discover all available Otel components together with some visibility into their quality.

## Explanation

The [OpenTelemetry Registry](https://opentelemetry.io/registry/) serves as a central catalogue of all known Otel components,
both provided by core maintainers of the project and any third party.

In order for a component to be included into Registry its authors have to fill a [self-assessment form](#registry-self-assessment-form).

The registry should allow a clear visibility of component's ownership, quality and compatibility with other Otel components.

A component can be removed from the Registry if any declaration from the self-assessment form is violated and not remedied
in a timely manner, provisionally one or two weeks.

## Internal details

We distinguish the following sources of OpenTelemetry components and integrations.

### Native or built-in instrumentations

Any library or framework can use the OpenTelemetry API to natively produce telemetry.

> We encourage library authors to submit their library for inclusion into the OpenTelemetry Registry.

Author: I think that only library authors themselves should make any claims about the integration when filling self-assessment form. So I don't see "on behalf of" model her, but I don't want to object to it neither.

> Do we expect authors to be incentivized to be included into registry

Yes, Registry will be one more channel to get visibility and feedback, hopefully

### Core components

Otel SIGs may provide instrumentation for any libraries that are deemed important enough by the SIG’s maintainers.
By doing this the SIG maintainers commit to support this instrumentation (including future versions of the library),
provide updates/fixes, including security patches and guarantee that they comply with Otel semantic conventions and best practices.

Depending on the SIG, these instances of core instrumentation may share the repository, infrastructure and maintainers with
the OpenTelemetry API/SDK implementation for this language or be separate.

All instances of core instrumentation must be included into the Otel Registry following the usual process described above.

### Contrib components

> Any language SIG may have one or more “contrib” repos containing components contributed by external developers.

Reviewer 7: ```suggestion
Any language SIG may have one or more “contrib” repos containing components contributed by developers with an interest in specific parts of the instrumentation ecosystem.
```

> Any language SIG may have one or more “contrib” repos containing components contributed by external developers.

Author: "not approvers of maintainers of this repository"

> These repositories may have a separate set of approvers/maintainers than the core API/SDK repo.

Author: I like your second suggestion below more. :) I propose rejecting this one and accepting the one below

SIG maintainers are encouraged to have a simplified process to promote contributors to approver/maintainer role in this repository.
A contrib repository may leverage the CODEOWNERS functionality of GitHub to assign maintainers to individual packages
even if this means granting write permissions to the whole repo.
The goal should be to distribute the load of reviewing PRs and accepting changes as much as possible.

All components in a contrib repository are expected to be included into the Otel Registry following the usual process described above.

> It should be very easy for external contributors to include their components into a contrib repo as opposed to hosting them separately.

Reviewer 8: > as opposed to hosting them separately

What do we expect to simplify for contributors? Security scanning and testing below is not a necessity (it is for us, but may not be for contributor). I think the intent of this sentence is different from how it may be interpreted. It may be harder as we likely will ask for things like build definitions, proper readme files, etc. which may not be the same in separate externally owned repo.

Maybe rephrase it emphasizing that maintainers of contrib repo should be open for adding new components

Author: @SergeyKanzhelev I tried to rephrase this, PTAL.

> It should be very easy for external contributors to include their components into a contrib repo as opposed to hosting them separately.

Author: There is no more "external contributors" in this OTEP.

This way they can reuse existing infrastructure for testing, publishing, security scanning etc.
This will also greatly simplify responsibility transfer between different maintainers if their priorities change. It also promotes the development

> and maintenance of a single instrumentation package for each instrumentation source, so that work isn't spread amongst multiple parallel solutions.

Reviewer 8: So if two developers came up with the different models to instrument a library, and do not want to collaborate. Or even worse, have slightly different semantics of attributes. Do we include both? Who decides on which wins?

Author: Most probably one of the integrations will be the first and already included into contrib repo. Then we first ask the second developer to try and collaborate with the first one. If this fails, then the second developer is free to host their vision in their own repo. Both integrations may be included into the Registry.

Not an ideal situation, but I don't have any better answer.

Reviewer 6: >  then the second developer is free to host their vision in their own repo. Both integrations may be included into the Registry.

I have seen this in the past. Indeed, there's no perfect solution, we can only mention our advise for such case.

Language SIGs are encouraged to provide a testing harness to verify that component adheres to OpenTelemetry semantic conventions
and [recommendations for Otel instrumentations design](https://docs.google.com/document/d/1YNRCg9fdjJgZRs56vvf7rfFPk06mhp781sWHYypOaAk/edit#)

> when OpenTelemetry starts publishing them.

Reviewer 7: I am eager to see support for testing OpenTelemetry instrumentation be included in the core of OTel packages, as a commitment to ecosystem health. This is not the same as a harness for semantic conventions, it's a statement that OTel APIs SHOULD offer testing support to make it easy to assert correct instrumentation.

Author: What does "correct" mean here apart from telemetry shape and form?

Reviewer 6: @jmacd We have testbed examples in a few scenario, where we try to test out both common and exotic scenarios of instrumentation patterns. Would that be enough? Or do you have something else in mind?

only core and contrib components can be published under "opentelemetry" namespace.
Native and external components are to be published under their own namespace.

In case the Otel SIG provides any kind of "all-in-one" instrumentation distribution (e.g. as Java and .NET do)
their should be an option to get a distribution with only core and contrib packages included.
The OpenTelemetry Registry should provide a way to easily obtain a list of these components.
The SIG may provide other distributions as well.
If possible, SIGs should provide a mechanism to include any external component during target application's build- or runtime.
This may mean a separate language-specific component API that all components are encouraged to implement.

## Trade-offs and mitigations

* How easy it is to get merge permission in contrib repo?
The harder it is, the larger is the maintenance burden on the core team.
The easier it is, the more uncertainty there is about the quality of contributions.
Can every language SIG decide this for themselves or should we decide together?

## Open questions

### Registry self-assessment form

> The exact list should be developed separately, but at least the component's author should declare that

Reviewer 8: maybe also include the "snippet description", "readme" and "documentation link"

* It uses permissive OSS license approved by CNCF
* It does not have any known security vulnerabilities
* It produces telemetry which adheres to OpenTelemetry semantic conventions
* If the Otel/SIG provides a testing harness to verify produced telemetry, that tests were used and passed
* Authors commit a reasonable effort into future maintenance of this component

### Component information in the Registry

The exact UI of component's page in the Registry is outside of this OTEP's scope, but some suggestions are:

* Link to the original repository of this component
* Clear indication of component's authors/maintainers and its Otel status (native/core/contrib/external)

> * Licence used

Reviewer 8: ```suggestion
* License used
```

Is it important for the registry? Maybe a better option for consumer to go to the source and check the license there. What if license changed at source repository, but nobody updated the registry? Are we liable as OpenTelemetry?

Author: This can be discussed in depth while implementing this OTEP in the Registry. This list here is certainly not a definitive one.

* Language and targeted library

> * Presence or absence of security vulnerabilities

Reviewer 8: maybe just link to the list of known issues? Or this suppose to be provided by a bot?

including integrations with various libraries and frameworks in all languages supported by OpenTelemetry.
We cannot possibly expect that all these integrations will be provided by the core maintainers of OpenTelemetry.
We hope that wider community will integrate their projects with OpenTelemetry.
We have to encourage that by providing great documentation, examples and tooling to integration authors,
while still providing our end-users with some way to discover all available OpenTelemetry components together with some visibility into their quality.

## Explanation

The [OpenTelemetry Registry](https://opentelemetry.io/registry/) serves as a central catalogue of all known OpenTelemetry components,
both provided by core maintainers of the project and any third party.

In order for a component to be included into Registry its authors have to fill a [self-assessment form](#registry-self-assessment-form).

The registry should allow a clear visibility of component's ownership, quality and compatibility with other OpenTelemetry components.

A component can be removed from the Registry if any declaration from the self-assessment form is violated and not remedied

> in a timely manner, provisionally one or two weeks.

Reviewer 9: 1-2 weeks is actually not a lot of time. A lot of open-source work is not done as part of someone's day job, and I worry that this would be a high-pressure tactic on folks who are contributing in their spare time, for no compensation. With a deadline of 1-2 weeks, it feels plausible that people may not even see the notification for a few days and thus significantly shorten their window to respond, not even counting the time it would take to actually remedy any deficiencies.

I would suggest 4 weeks as a starting point (maybe a little more, maybe a little less). That should give folks plenty of time to see the notice (and clarify anything that needs), and then time to actually correct any problems.

Reviewer 6: Quite a valid point. I'd be up for 3 weeks instead. @iNikem any opinion on this?

Reviewer 9: For me, I think 3 weeks is much more reasonable. And we can always revisit the length of time later based on maintainer feedback. 😄

Reviewer 2: all of this can be decided when this proposal becomes part of the specification, correct?

Author: These 1-2 weeks here are "provisional", meaning they are meant to indicate a rough estimate. E.g. "not 1 day and not 2 months, somewhere in between"

Reviewer 9: I think it's also fine to resolve if/when merged into the specification; however it may be useful to our future selves to have this information in the OTEP now rather than later. Specifically if `1-2 weeks` means `not 1 day and not 2 months, somewhere in between` then that might be something other readers cannot know without finding this specific comment thread.

Reviewer 5: Conversion of OTEP into specification was meant to be more of a mechanical process, not another round of debates. I would recommend resolving this issue now.

I agree 2 weeks is too short. It actually does not need to be that short for the specific goals, because a warning can be added to the registry much faster to alert people of an issue with some contrib module, and even override the specific item in self-assessment.

Author: Increased to "3 to 4"

## Internal details

We distinguish the following sources of OpenTelemetry components and integrations.

### Native or built-in instrumentations

Any library or framework can use the OpenTelemetry API to natively produce telemetry.
We encourage library authors to submit their library for inclusion into the OpenTelemetry Registry.

### Core components

> Otel SIGs may provide instrumentation for any libraries that are deemed important enough by the SIG’s maintainers.

Author: Replaced with "OpenTelemetry"
